http/events.go

```diff
@@ -69,14 +87,80 @@ func handleEventsSubscribeWebsocket(args eventSubscribeArgs) (websocket.StatusCo
 				logger.Warn("Could not serialize websocket event", "error", err)
 				return 0, "", err
 			}
-			err = args.conn.Write(ctx, messageType, messageBytes)
+			err = sub.conn.Write(ctx, messageType, messageBytes)
 			if err != nil {
 				return 0, "", err
 			}
 		}
 	}
 }
 
+// allowMessageCached checks that the message is allowed to received by the websocket.
+// It caches results for specific namespaces, data paths, and event types.
+func (sub *eventSubscriber) allowMessageCached(message *logical.EventReceived) bool {
+	// default to the subscribe endpoint if there is no specific data path
+	messageNs := strings.Trim(message.Namespace, "/")
+	dataPath := "sys/events/subscribe"
+	if message.Event.Metadata != nil {
+		dataPathField := message.Event.Metadata.GetFields()[logical.EventMetadataDataPath]
+		if dataPathField != nil {
+			dataPath = dataPathField.GetStringValue()
+		}
+	}
+	cacheKey := fmt.Sprintf("%v!%v!%v", messageNs, dataPath, message.EventType)
```

Comment:
I'm a little confused about how the cache works here -- let's say I update my KV three times, wouldn't that have the same NS, path, and event type?

Reply:
Exactly. If you just did a KV update to a secret and we generated an event and checked that it is allowed, and then we saw another event right after for the same secret, we wouldn't want to do the whole ACL check again, so we cache the result (for 5 minutes).

Reply:
Ahh, that makes complete sense. Thanks for clarifying!